src/wallet/rpcwallet.cpp

```diff
@@ -3956,6 +3956,84 @@ void fill_psbt(const CWallet* pwallet, PartiallySignedTransaction& psbtx, const
             psbtx.inputs.push_back(input);
         }
     }
+
+    // Fill in the bip32 keypaths and redeemscripts for the outputs so that hardware wallets can identify change
+    if (include_output_info) {
+        for (const CTxOut& out : txConst->vout) {
+            // Get scriptpubkey and check for redeemScript or witnessscript
+            txnouttype type;
+            std::vector<std::vector<unsigned char>> solns;
+            Solver(out.scriptPubKey, type, solns);
+            // Get script hashes
+            if (type == TX_SCRIPTHASH) {
+                // get the hash and find it in the wallet.
+                CScript redeem_script;
+                uint160 hash(solns[0]);
+                if (!pwallet->GetCScript(CScriptID(hash), redeem_script)) {
+                    // We don't have this script, skip it
+                    continue;
+                }
+
+                // put redeem_script in map
+                psbtx.redeem_scripts.emplace(hash, redeem_script);
+
+                // Now check whether the redeem_script is a witness script
+                solns.clear();
+                Solver(redeem_script, type, solns);
+            }
+            // Get witness scripts
+            if (type == TX_WITNESS_V0_SCRIPTHASH) {
+                // Get the hash from the solver return
+                uint160 hash;
+                CRIPEMD160().Write(&solns[0][0], solns[0].size()).Finalize(hash.begin());
+
+                // Lookup hash from wallet
+                CScript witness_script;
+                if (!pwallet->GetCScript(CScriptID(hash), witness_script)) {
+                    // We don't have this script, skip it
+                    continue;
+                }
+
+                // Put witness script in map
+                uint256 hash256(solns[0]);
+                psbtx.witness_scripts.emplace(hash256, witness_script);
+
+                // Decode the witness script
+                solns.clear();
+                Solver(witness_script, type, solns);
+            }
+            // Get public keys if hd is enabled
+            if (pwallet->IsHDEnabled()) {
+                if (type == TX_PUBKEYHASH || type == TX_WITNESS_V0_KEYHASH) {
+                    uint160 hash(solns[0]);
+                    CKeyID keyID(hash);
+                    if (!pwallet->HaveKey(keyID)) {
```

Comment:
What we really care about is that we have the hdkeypath, not particularly that the key is present or not. I don't think these `HaveKey` checks are necessary since `add_keypath_to_map` will deal with that case specifically.

Reply:
Right, fixed.